Allocate ipcache mappings for service backends

joestringer opened the pull request:
```release-note
Fix regression that caused policies with `ToServices` rules to not allow traffic to services with external backends
```

If adding a policy failed, then we previously would release the                                                                                                                          
identities that were allocated, but leave the ipcache mappings in the           
kvstore for those CIDR->ID pairs. Release the ipcache mappings as well.         
                                                                                
If some of the identities cannot be located in the kvstore, then none of        
the identities or ipcache entries would be deleted when removing a              
policy. Fix this up to instead attempt to locate as many of the entries         
as possible, then free as many as can be located.                               
                                                                                
When generating CIDR rules in k8s translation, allocate identities for          
the new CIDRs (for service backends) and create CIDR->Identity mappings         
in the IPCache.                                                                 
                                                                                
Fixes: 4692732a789f ("daemon: Allocate identities for CIDRs")                   
Fixes: #4570

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4587)
<!-- Reviewable:end -->

joestringer committed:
k8s: Allocate ipcache mappings for service backends

When generating CIDR rules in k8s translation, allocate identities for
the new CIDRs (for service backends) and create CIDR->Identity mappings
in the IPCache.

Fixes: #4570

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Refactor getting CIDR prefixes from CIDR strings

Put this code into the ip package so that an upcoming commit can reuse
from common/.

Signed-off-by: Joe Stringer <joe@covalent.io>

raybejjani commented on daemon/policy.go:
```diff
@@ -237,28 +235,20 @@ func (d *Daemon) PolicyAdd(rules api.Rules, opts *AddOptions) (uint64, error) {
 		// Don't leak identities allocated above.
-		identity.ReleaseSlice(prefixIdentities)
-
+		if err2 := ipcache.ReleaseCIDRs(prefixes); err2 != nil {
```
This was the earlier behaviour, I think, but if we already had some CIDRs added, and we see a failure for a new set of CIDRs that include those older CIDRs, this would then remove those as well, right?

raybejjani replied:
Also, doesn't AllocateCIDRs do the release for us on error?

raybejjani replied:
Oh, I guess there is some local/remote difference that is too complicated for me to understand.

joestringer replied:
Hmm. In this case all of the CIDRs for this policy were successfully added to the kvstore, but then during import of the policy into the repository it was rejected so this should now roll back all of the kvstore interactions - both for the identitie allocations (which was previously done) and for the ipcache mappings (which was previously skipped).

We're using the same slice of prefixes here as were used during allocation, so I'd expect this to release the identities and mappings that were just created in the kvstore.

With regards to interactions with prefixes other CIDR policies, the theory is that each CIDR rule takes its own reference on each kvstore key and the delete only occurs in the kvstore if all of them are deleted. That said I suspect there's another bug in that area with how the ipcache is managed that means this is not quite the case. I plan to follow up on that separately, as I think there are ways to trigger via other means - that is, addition of two CIDR rules followed by removal of one.

joestringer replied:
Here's the PR: #4611

raybejjani commented on pkg/k8s/rule_translate.go:
```diff
@@ -160,11 +182,23 @@ func deleteToCidrFromEndpoint(
+		prefixes := policy.GetPrefixesFromCIDRSet(deleted)
+		if err := ipcache.ReleaseCIDRs(prefixes); err != nil {
+			// Don't pass this up to the caller, as it would prevent the
+			// deletion of other services or the setup of new ones!
```
Perhaps mention this error suppression in the docstring above?

rlenglet commented on common/types/loadbalancer.go:
```diff
@@ -214,6 +215,19 @@ func NewK8sServiceEndpoint() *K8sServiceEndpoint {
+	}
+	valid, invalid := ip.GetPrefixesFromCIDR(prefixes)
+	if len(invalid) > 0 {
+		return nil, fmt.Errorf("Invalid IPs specified as backends: %+v", invalid)
```
`Invalid` -> `invalid`

rlenglet commented on daemon/policy.go:
```diff
@@ -237,28 +235,20 @@ func (d *Daemon) PolicyAdd(rules api.Rules, opts *AddOptions) (uint64, error) {
-		identity.ReleaseSlice(prefixIdentities)
-
+		if err2 := ipcache.ReleaseCIDRs(prefixes); err2 != nil {
+			log.WithError(err2).WithField("prefixes", prefixes).Debug(
```
Debug -> Warn (?)

rlenglet commented on pkg/ip/cidr.go:
```diff
@@ -0,0 +1,51 @@
+
+// GetPrefixesFromCIDR fetches all CIDRs referred to by the specified slice
+// and returns them as regular golang CIDR objects.
+func GetPrefixesFromCIDR(cidrs []string) (valid []*net.IPNet, invalid []string) {
```
`ParseCIDRs` (?)

rlenglet commented on pkg/ipcache/cidr.go:
```diff
@@ -0,0 +1,86 @@
+		if err2 := identity.ReleaseSlice(prefixIdentities); err2 != nil {
+			log.WithError(err2).WithFields(logrus.Fields{
+				"identities": prefixIdentities,
+			}).Debug("Failed during CIDR allocation")
```
Debug -> Warn

rlenglet commented on pkg/ipcache/cidr.go:
```diff
@@ -0,0 +1,86 @@
+			err = err2
+		}
+		scopedLog.WithError(err2).Debug(
+			"Could not locate identities for CIDRs during release")
```
locate -> find

rlenglet commented on pkg/ipcache/cidr.go:
```diff
@@ -0,0 +1,86 @@
+				err = err2
+			}
+			log.WithError(err2).WithFields(logrus.Fields{
+				"identities": prefixIdentities,
```
This field key is used in many places. Define it in `logfields`.

joestringer committed:
daemon: Release ipcache entries on policy add failure

If adding a policy failed, then we previously would release the
identities that were allocated, but leave the ipcache mappings in the
kvstore for those CIDR->ID pairs. Release the ipcache mappings as well.

Fixes: 4692732 ("daemon: Allocate identities for CIDRs")

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Release identities in failure condition

If some of the identities cannot be located in the kvstore, then none of
the identities or ipcache entries would be deleted when removing a
policy. Fix this up to instead attempt to locate as many of the entries
as possible, then free as many as can be located.

Fixes: 4692732 ("daemon: Allocate identities for CIDRs")

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Refactor ipcache CIDR allocations to ipcache

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
ipcache: Improve comments for reference counter

Clean up some minor typos in the comments.

Signed-off-by: Joe Stringer <joe@covalent.io>
